CAN-MCAN line 1 ISR - add more debug information

KozhinovAlexander opened the pull request:
In some cases CAN-Bus communication may be disturbed by other parts of SoC. For example USB may disturb CAN communication if electircal wiring of the board is not done properly or the system is in development stage.
This change adds debug-prints s.t. CAN-Bus protocol errors in arbitration or data phases can be
detected while in CAN-Bus ISR.

In my case I am getting 16 ISR calls with 16 same CAN messages as soon as I enable USB and CAN on my nucleo_h745zi_q board with waveshare-can-rs485 shield. Currently I am investigating the problem. But it shall work, cause there are plenty of STM32 designs with CAN and USB enabled.

KozhinovAlexander committed:
drivers: can_mcan: can_mcan_line_1_isr: add more debug outputs

In some cases CAN-Bus communication may be
disturbed by other parts of SoC. For example USB
may disturb CAN communication if electircal wiring
of the board is not done properly or the system is
in development stage.
This change adds debug-prints s.t. CAN-Bus protocol
errors in arbitration or data phases can be
detected while in CAN-Bus ISR.

Signed-off-by: Alexander Kozhinov <ak.alexander.kozhinov@gmail.com>

henrikbrixandersen commented:
> 1. [ebbb155](https://github.com/zephyrproject-rtos/zephyr/pull/73544/commits/ebbb15527a75940b831b2a4d1eabc0e5d5b90073) Reduce number of `can_mcan_read_reg()` function calls in ISR

What is the purpose?

KozhinovAlexander commented:
> > 1. [ebbb155](https://github.com/zephyrproject-rtos/zephyr/pull/73544/commits/ebbb15527a75940b831b2a4d1eabc0e5d5b90073) Reduce number of `can_mcan_read_reg()` function calls in ISR
> 
> What is the purpose? You do not reduce the number of function calls, you will always need to read it once per loop iteration?

Thank you Henrik for you review.

The main purpose of this PR is to add debug printout s.t. developers may be able to find possible issues faster.
At the moment on my nucleo_h745zi_q board I have a problem, that I am receiving 16 can messages instead of single one as soon as CAN and USB are enabled together. I am getting IR register value of 0x800001. Separately USB an CAN works totally fine. My idea is, that here some electrical layout or supply capacitors issues in my board, which I am currently investigating. Exactly for that purposes mentioned debug printouts may be really helpful.
Second change with reducing number of IR register readings is optional in my opinion. Let's, say we save maybe few machine cycles, which may be nice in an ISR.
You're absolutely right - there is no reduction of IR register reading in the loop itself. But if you take a look at the line 800, there is not IR register read anymore. Thus one additional call were eliminated.
I hope it helps you with further. And of course you're free to drop this PR, if you do not see enough value. But I also believe, that my changes may help a lot of developers.

henrikbrixandersen commented:
Thanks you for that explanation. I was mainly confused by the title of the PR compared with the actual changes. 

If you could update the title/descrption/Git commit logs to be a bit more elaborate, I'm fine with this change. Thanks!

KozhinovAlexander commented:
> Thanks you for that explanation. I was mainly confused by the title of the PR compared with the actual changes.
> 
> If you could update the title/descrption/Git commit logs to be a bit more elaborate, I'm fine with this change. Thanks!

Thank you for very fast reponse. I've rebased this PR over newest master, changed PR name and updated descriptions in both commits. Please take a look again. I hope it fits now. Please let me know, if not.

martinjaeger commented on drivers/can/can_mcan.c:
```diff
@@ -797,12 +797,24 @@ void can_mcan_line_1_isr(const struct device *dev)
-		return;
-	}
+	do {
+		err = can_mcan_read_reg(dev, CAN_MCAN_IR, &ir);
```
Maybe I'm missing something, but I don't see how this reduces the number of calls to `can_mcan_read_reg`. It is still called at least once and afterwards as long as the condition is not met.

However, with the `do { ... } while (true)` loop it's not completely obvious anymore what the stop condition is. I think I prefer keeping the existing implementation.

KozhinovAlexander replied:
Thank you for the fast response. Basically my change avoids two IR register reads in case while-loop will be approached. 
Thus you got it right. 
I agree - this change is mainly cosmetic, but shall still saves few machine-cycles, which may be useful on very low budget MCUs with low speed.
I would suggest also Henrik to vote on this change. And of course I will remove it, if you both vote against. I am fine with it.

henrikbrixandersen replied:
I must admit that I too see no real issue with the current implementation and find it more readable (but I may be biased).

alexanderwachter replied:
I too think the current implementation is more straight forward because there is a single break condition.

KozhinovAlexander replied:
> I too think the current implementation is more straight forward because there is a single break condition.
 
Hello Alexander and thank you for your review. Well, in my implementation there is also only one break-condition, cause `while(true)` is not a break-condition ;) 
I see, majority voting against while-loop redesign. I will revert this changes and leave only debug priontouts.

KozhinovAlexander commented:
Dear reviewers I've done changes as it was requested by you. In particular:

- while-loop redesign with IR register reds reduction were dropped.
- description and title of this PR were updated.

Now I hpe the changes are more stragihtforward and could be evaluated by you again. Thanks a lot.

KozhinovAlexander commented:
Found my problem: The value of sjw can bus parameter were equal to 1.
Added debug info helped me to find the problem faster. :)
